File: apps/web/test/unit/application/use_cases/toggle_todo_use_case.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";
import { ToggleTodoUseCase } from "../../../../src/core/application/use_cases/toggle_todo_use_case";
import type { TodoEntity } from "../../../../src/core/domain/entities/todo_entity";
import type { TodoRepository } from "../../../../src/core/domain/repositories/todo_repository";
import { Priority } from "../../../../src/core/domain/value_objects/priority";

describe("ToggleTodoUseCase", () => {
  let mockRepository: TodoRepository;
  let useCase: ToggleTodoUseCase;

  beforeEach(() => {
    mockRepository = {
      findAll: vi.fn(),
      findById: vi.fn(),
      create: vi.fn(),
      update: vi.fn(),
      delete: vi.fn(),
    };
    useCase = new ToggleTodoUseCase(mockRepository);
  });

  it("should toggle todo from incomplete to completed", async () => {
    const existing: TodoEntity = {
      id: "test-id",
      title: "Test Todo",
      priority: Priority.MEDIUM,
      completed: false,
      createdAt: "2024-01-01T00:00:00.000Z",
      updatedAt: "2024-01-01T00:00:00.000Z",
    };

    const updated: TodoEntity = {
      ...existing,
      completed: true,
      updatedAt: "2024-01-02T00:00:00.000Z",
    };

    vi.mocked(mockRepository.findById).mockResolvedValue(existing);
    vi.mocked(mockRepository.update).mockResolvedValue(updated);

    const result = await useCase.execute("test-id");

    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data.completed).toBe(true);
      expect(mockRepository.update).toHaveBeenCalled();
    }
  });

  it("should toggle todo from completed to incomplete", async () => {
    const existing: TodoEntity = {
      id: "test-id",
      title: "Test Todo",
      priority: Priority.MEDIUM,
      completed: true,
      createdAt: "2024-01-01T00:00:00.000Z",
      updatedAt: "2024-01-01T00:00:00.000Z",
    };

    const updated: TodoEntity = {
      ...existing,
      completed: false,
      updatedAt: "2024-01-02T00:00:00.000Z",
    };

    vi.mocked(mockRepository.findById).mockResolvedValue(existing);
    vi.mocked(mockRepository.update).mockResolvedValue(updated);

    const result = await useCase.execute("test-id");

    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data.completed).toBe(false);
    }
  });

  it("should fail when todo not found", async () => {
    vi.mocked(mockRepository.findById).mockResolvedValue(null);

    const result = await useCase.execute("non-existent");

    expect(result.success).toBe(false);
    if (!result.success) {
      expect(result.error).toBe("Todo not found");
    }
  });
});

File: apps/mobile/android/app/src/test/java/com/erluxman/productive/core/application/use_cases/ToggleTodoUseCaseTest.kt
package com.erluxman.productive.core.application.use_cases

import com.erluxman.productive.core.domain.entities.TodoEntity
import com.erluxman.productive.core.domain.repositories.TodoRepository
import com.erluxman.productive.test_utils.TestFixtures
import io.mockk.coEvery
import io.mockk.mockk
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runTest
import org.junit.Assert.*
import org.junit.Test

@OptIn(ExperimentalCoroutinesApi::class)
class ToggleTodoUseCaseTest {
    
    private val repository: TodoRepository = mockk()
    private val useCase = ToggleTodoUseCase(repository)
    
    @Test
    fun `execute should toggle todo from incomplete to complete`() = runTest {
        // Given
        val todo = TestFixtures.createTodoEntity(id = "1", isCompleted = false)
        val toggledTodo = todo.toggleCompletion()
        coEvery { repository.getById("1") } returns Result.success(todo)
        coEvery { repository.update(toggledTodo) } returns Result.success(toggledTodo)
        
        // When
        val result = useCase.execute("1")
        
        // Then
        assertTrue(result.isSuccess)
        result.fold(
            onSuccess = { updatedTodo ->
                assertTrue(updatedTodo.isCompleted)
            },
            onFailure = { fail("Should not fail") }
        )
    }
    
    @Test
    fun `execute should toggle todo from complete to incomplete`() = runTest {
        // Given
        val todo = TestFixtures.createTodoEntity(id = "1", isCompleted = true)
        val toggledTodo = todo.toggleCompletion()
        coEvery { repository.getById("1") } returns Result.success(todo)
        coEvery { repository.update(toggledTodo) } returns Result.success(toggledTodo)
        
        // When
        val result = useCase.execute("1")
        
        // Then
        assertTrue(result.isSuccess)
        result.fold(
            onSuccess = { updatedTodo ->
                assertFalse(updatedTodo.isCompleted)
            },
            onFailure = { fail("Should not fail") }
        )
    }
    
    @Test
    fun `execute should return failure when id is blank`() = runTest {
        // When
        val result = useCase.execute("")
        
        // Then
        assertTrue(result.isFailure)
        result.fold(
            onSuccess = { fail("Should not succeed") },
            onFailure = { error ->
                assertTrue(error is IllegalArgumentException)
                assertEquals("Todo ID cannot be empty", error.message)
            }
        )
    }
    
    @Test
    fun `execute should return failure when todo not found`() = runTest {
        // Given
        val exception = Exception("Todo not found")
        coEvery { repository.getById("1") } returns Result.failure(exception)
        
        // When
        val result = useCase.execute("1")
        
        // Then
        assertTrue(result.isFailure)
        result.fold(
            onSuccess = { fail("Should not succeed") },
            onFailure = { error ->
                assertEquals(exception, error)
            }
        )
    }
}

